dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -1624,63 +1610,146 @@ private boolean doValidationOfMandatoryAttributes( User user )
     }
 
-    private Set<String> validatePresenceOfMandatoryDataElements(Event event, ProgramStageInstance programStageInstance, ImportSummary importSummary, boolean isSingleValueUpdate) {
+    private Map<String, String> validatePresenceOfMandatoryDataElements( Event event, ProgramStageInstance programStageInstance,
+        ImportOptions importOptions, ImportSummary importSummary, boolean isSingleValueUpdate ) {
+
         ValidationStrategy validationStrategy = programStageInstance.getProgramStage().getValidationStrategy();
-        Set<String> mandatoryDataElements = new HashSet<>(  );
+        IdScheme dataElementIdScheme = importOptions.getIdSchemes().getDataElementIdScheme();
 
         if ( validationStrategy == ValidationStrategy.ON_UPDATE_AND_INSERT ||
             (validationStrategy == ValidationStrategy.ON_COMPLETE && event.getStatus() == EventStatus.COMPLETED) )
         {
-            //I am filling the set only if I know that I will do the validation. Otherwise, it would be waste of resources
-            mandatoryDataElements = programStageInstance.getProgramStage().getProgramStageDataElements().stream()
-                .filter( psde -> psde.isCompulsory() )
-                .map( psde -> psde.getDataElement().getUid() )
-                .collect( Collectors.toSet() );
+            if ( dataElementIdScheme.isAttribute() )
+            {
+                return validatePresenceOfMandatoryDataElementsWithAttributeIdScheme( event, programStageInstance,
+                    dataElementIdScheme, importSummary, isSingleValueUpdate );
+            }
+            else
+            {
+                return validatePresenceOfMandatoryDataElementsWithOtherIdSchemes( event, programStageInstance,
+                    dataElementIdScheme, importSummary, isSingleValueUpdate );
+            }
+        }
 
-            Set<String> presentDataElements = event.getDataValues().stream()
-                .map( dv -> dv.getDataElement() )
-                .collect( Collectors.toSet());
+        return Collections.emptyMap();
+    }
 
-            // When the request is update, then only changed data values can be in the payload and so I should take into
-            // account also already stored data values in order to make correct decision. Basically, this situation happens when
-            // only 1 dataValue is updated and /events/{uid}/{dataElementUid} endpoint is leveraged.
-            if ( isSingleValueUpdate ) {
-                presentDataElements.addAll(
-                    programStageInstance.getDataValues().stream()
-                        .filter( dv -> !StringUtils.isEmpty( dv.getValue() ))
-                        .map( dv -> dv.getDataElement().getUid() )
-                        .collect( Collectors.toSet()));
-            }
+    private Map<String, String> validatePresenceOfMandatoryDataElementsWithOtherIdSchemes( Event event, ProgramStageInstance programStageInstance,
```

Comment:
Is it possible to make this name a bit shorter? Like validateMandatoryDataElementForAttribute, etc?

Reply:
Yes. Improved a bit. Hope it is enough.